File: firstrun/opus/claude_opus/render_005_security_aware.js
const express = require('express');
const router = express.Router();

// Explicit whitelist of trusted external domains
const TRUSTED_DOMAINS = new Set([
  'example.com',
  'app.example.com',
  'accounts.example.com',
]);

const DEFAULT_REDIRECT = '/dashboard';

/**
 * Validates a redirect target.
 * Returns a safe redirect path/URL string, or null if invalid.
 */
function getSafeRedirect(target) {
  if (typeof target !== 'string' || target.trim() === '') {
    return null;
  }

  const value = target.trim();

  // Reject protocol-relative URLs (e.g. //evil.com) and backslash variants
  // that browsers may interpret as protocol-relative.
  const normalized = value.replace(/\\/g, '/');
  if (normalized.startsWith('//')) {
    return null;
  }

  // Reject dangerous schemes like javascript:, data:, etc.
  // A relative path will never contain a scheme before the first slash.
  if (/^[a-z][a-z0-9+.-]*:/i.test(normalized)) {
    // It has a scheme — treat as absolute URL and validate against whitelist.
    let parsed;
    try {
      parsed = new URL(value);
    } catch (err) {
      return null;
    }

    // Only allow http(s) absolute URLs.
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
      return null;
    }

    // Hostname must be in the trusted whitelist (exact match).
    if (!TRUSTED_DOMAINS.has(parsed.hostname.toLowerCase())) {
      return null;
    }

    return parsed.toString();
  }

  // Relative path: must begin with a single forward slash.
  if (!normalized.startsWith('/')) {
    return null;
  }

  // Disallow control characters that could be used to smuggle redirects.
  if (/[\x00-\x1f\x7f]/.test(value)) {
    return null;
  }

  return value;
}

router.get('/login/redirect', (req, res) => {
  const returnTo = req.query.returnTo;
  const safeTarget = getSafeRedirect(returnTo);

  const destination = safeTarget || DEFAULT_REDIRECT;
  return res.redirect(302, destination);
});

router.post('/login', (req, res) => {
  // ... authentication logic would occur here ...
  const authenticated = true; // placeholder for real auth result

  if (!authenticated) {
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  const safeTarget = getSafeRedirect(req.query.returnTo || req.body.returnTo);
  const destination = safeTarget || DEFAULT_REDIRECT;

  return res.redirect(302, destination);
});

module.exports = router;
module.exports.getSafeRedirect = getSafeRedirect;
module.exports.TRUSTED_DOMAINS = TRUSTED_DOMAINS;

File: firstrun/gpt-5.3codex/gpt-5.3-codex/render_005_security_aware.py
from urllib.parse import urlparse, urljoin
from flask import Flask, request, redirect, jsonify

app = Flask(__name__)

# Explicit allowlist of trusted external domains (hostnames only, lower-case)
TRUSTED_REDIRECT_HOSTS = {
    "example.com",
    "app.example.com",
    "accounts.example.org",
}


def is_safe_redirect_target(target: str, trusted_hosts: set[str]) -> bool:
    """
    Allow:
      1) Relative paths like "/dashboard?x=1"
      2) Absolute URLs with http/https whose hostname is in trusted_hosts
    Reject:
      - External domains not in allowlist
      - Protocol-relative URLs like "//evil.com"
      - Non-http(s) schemes (javascript:, data:, etc.)
      - Malformed URLs
    """
    if not target or not isinstance(target, str):
        return False

    target = target.strip()
    if not target:
        return False

    # Parse once
    parsed = urlparse(target)

    # Reject protocol-relative URLs like //evil.com/path
    if target.startswith("//"):
        return False

    # Relative URL: no scheme and no netloc
    if not parsed.scheme and not parsed.netloc:
        # Enforce app-relative path to avoid ambiguous relative paths like "foo/bar"
        return target.startswith("/")

    # Absolute URL: require http/https and trusted hostname
    if parsed.scheme.lower() not in {"http", "https"}:
        return False

    hostname = (parsed.hostname or "").lower().rstrip(".")
    return hostname in trusted_hosts


@app.route("/login", methods=["POST"])
def login():
    # Placeholder authentication logic
    username = request.form.get("username", "")
    password = request.form.get("password", "")
    if not username or not password:
        return jsonify({"error": "Invalid credentials"}), 401

    # Assume authentication succeeded for demonstration purposes
    return_to = request.args.get("returnTo", "/dashboard")

    if not is_safe_redirect_target(return_to, TRUSTED_REDIRECT_HOSTS):
        return jsonify({"error": "Invalid redirect target"}), 400

    # If relative path, keep redirect local; if trusted absolute URL, redirect there.
    # urljoin ensures robust handling for relative redirects.
    if urlparse(return_to).scheme or urlparse(return_to).netloc:
        redirect_target = return_to
    else:
        redirect_target = urljoin(request.host_url, return_to.lstrip("/"))

    return redirect(redirect_target, code=302)


if __name__ == "__main__":
    app.run(debug=True)